cpp/src_prims/sparse/batched/csr.h

```diff
@@ -0,0 +1,721 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <algorithm>
+#include <memory>
+#include <vector>
+
+#include <thrust/execution_policy.h>
+#include <thrust/for_each.h>
+#include <thrust/iterator/counting_iterator.h>
+
+#include <cuml/common/utils.hpp>
+#include <cuml/cuml.hpp>
+
+#include "linalg/batched/matrix.h"
+#include "linalg/cusolver_wrappers.h"
+#include "matrix/matrix.h"
+
+namespace MLCommon {
+namespace Sparse {
+namespace Batched {
+
+/**
+ * Kernel to construct batched CSR sparse matrices from batched dense matrices
+ * 
+ * @note The construction could coalesce writes to the values array if we
+ *       stored a mix of COO and CSR, but the performance gain is not
+ *       significant enough to justify complexifying the class.
+ * 
+ * @param[in]  dense      Batched dense matrices. Size: m * n * batch_size
+ * @param[in]  col_index  CSR column index.       Size: nnz
+ * @param[in]  row_index  CSR row index.          Size: m + 1
+ * @param[out] values     CSR values array.       Size: nnz * batch_size
+ * @param[in]  batch_size Number of matrices in the batch
+ * @param[in]  m          Number of rows per matrix
+ * @param[in]  n          Number of columns per matrix
+ * @param[in]  nnz        Number of non-zero elements in each matrix
+ */
+template <typename T>
+static __global__ void dense_to_csr_kernel(const T* dense, const int* col_index,
```

Comment:
Currently, this kernel is doing one batch per thread.
If the matrix dimensions are big, we are better off doing one batch per threadblock, right?

Reply:
I didn't really overthink this kernel because of it's insignificant cost (`O(nnz + m)`, called only once).

But I don't even think using one batch element per block would accelerate it. Here's why:
- The only way to do one batch element per block is to give different rows to each thread. The matrix being sparse, different rows will have different numbers of non-zero elements, so most threads will sit idle while the busiest row copies its values.
- We can't benefit from shared memory in this kernel so there isn't really any advantage of using one batch element per block (every data read only once, no good access pattern)
- Using one thread per batch element works well in our case of relatively large batches and relatively small matrices.

@teju85 do you agree with my arguments above?

Reply:
One way to do one batch element per block would be to store a COO representation of the sparse matrix because we could have one thread per nnz element, but the performance gain is not worth the trouble.

Reply:
I certainly get your points. However, the main reason I asked this question was because this operation is currently inside ml-prims and potentially other folks might end up using it. Thus, it'll be better to mention the perf characteristics of this kernel in its doxygen comments. That way, another fellow using this prim knows exactly what this prim does as well as performs.

Reply:
BTW, I'm ok keeping this kernel as is and only optimize it if and when this becomes a bottleneck.

Reply:
Ok, I added a few comments about performance.
Basically the whole implementation is designed for large batches of relatively small matrices, in other cases the dense counterparts would always perform better.